### ISSUE TITLE
[Android 14] Support partial image/video permission access

### ISSUE THREAD

**Issue by thomashorta** [wordpress-mobile/WordPress-Android#19016]
Android 14 introduced changes to how the image and video permissions work, making it possible for users to give only access to selected media. 

Some [changes](https://developer.android.com/about/versions/14/changes/partial-photo-video-access#effects-declaring) are needed to support it on apps running on Android 14, by implementing the new `READ_MEDIA_VISUAL_USER_SELECTED` permission and doing some extra changes when requesting image/video permissions.

Internal ref: pcdRpT-3Bg-p2

https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/ff676415-5218-42a3-a41e-9ff629b17249

##### Tested on Emulator, Android 14, JPAndroid / WPAndroid 23.0

```[tasklist]
### Tasks
- [ ] https://github.com/wordpress-mobile/WordPress-Android/pull/19103
- [ ] https://github.com/wordpress-mobile/WordPress-Android/pull/19109
- [ ] ~Implement `READ_MEDIA_VISUAL_USER_SELECTED` in Stories demo app~
- [ ] ~Implement `READ_MEDIA_VISUAL_USER_SELECTED` in FluxC demo app~
```

**Comment by fluiddot** [comment]
@thomashorta I assume the update of `compileSdkVersion` would be needed only in WP-Android and not in the libraries. Is this accurate?

**Comment by irfano** [comment]
> @thomashorta I assume the update of `compileSdkVersion` would be needed only in WP-Android and not in the libraries. Is this accurate?

In the migration process, I believe we should upgrade all repositories. But for this task, I think we should upgrade WPAndroid and stories. It is what I did in the Android 13 upgrade, https://github.com/wordpress-mobile/WordPress-Android/issues/17715.

**Comment by fluiddot** [comment]
> [..] But for this task, I think we should upgrade WPAndroid and stories. It is what I did in the Android 13 upgrade, https://github.com/wordpress-mobile/WordPress-Android/issues/17715.

I noticed that we also had to update `AztecEditor-Android` (https://github.com/wordpress-mobile/WordPress-Android/issues/17715#issuecomment-1480392100) during the Android 13 upgrade. Not sure if it would be needed for this case too 🤔.

**Comment by thomashorta** [comment]
@fluiddot @irfano 

I believe `compileSdkVersion` can be updated whenever we need to actually access new APIs to fix/implement something. As the name says, it's only used for compilation so it has no effect in runtime, since the code runs on whatever SDK the system has.

Initially, to support the partial image/video access I only really need to update the `compileSdkVersion` of `WordPress-Android` since it will need to access the new permission available in SDK 34. 

Do you think it's worth updating the `compileSdkVersion` of other libs as well, anyway? My personal opinion is that when we do the actual migration to Android 14 (changing `targetSdkVersion`) we could do the `compileSdkVersion` as well, wdyt?

**Comment by thomashorta** [comment]
I posted a comment but then re-read the comments above and deleted it. This should be accurate:
> In the migration process, I believe we should upgrade all repositories. But for this task, I think we should upgrade WPAndroid and stories. It is what I did in the Android 13 upgrade, https://github.com/wordpress-mobile/WordPress-Android/issues/17715.

I only really need to update the `compileSdkVersion` of `WP-Android` to make things work and should have PRs for it later today.

I was not aware the `stories`, `FluxC`, and `Aztec` demo Apps were using the permissions as well, so we might need to make changes there to support the new permission model for media. I will look into it later today.

Thanks for bringing that up! 🙇

**Comment by ParaskP7** [comment]
👋 @thomashorta !

> Do you think it's worth updating the compileSdkVersion of other libs as well, anyway? My personal opinion is that when we do the actual migration to Android 14 (changing targetSdkVersion) we could do the compileSdkVersion as well, wdyt?

Although you deleted this comment, my take and suggestion with relation to the `compileSdkVersion` upgrade on all other libs is that this is done in parallel, or at least at the same time the main `compileSdkVersion` upgrade on JP/WPAndroid is happening.

This is because doing the `compileSdkVersion` upgrade later and as part of the `targetSdkVersion` upgrade will delay this later by a significant factor. Also, doing the `compileSdkVersion` upgrade sooner than later helps us in terms of new Lint and deprecation warnings, which we could then start plan on dealing with a bit in advance.

**Comment by thomashorta** [comment]
> Although you deleted this comment, my take and suggestion with relation to the compileSdkVersion upgrade on all other libs is that this is done in parallel, or at least at the same time the main compileSdkVersion upgrade on JP/WPAndroid is happening.

In that case, would the course of action be like the following?
- Check all internal libs we use (from our `build.gradle` file)
- Update the compileSdkVersion of all of them to 34
- Fix any Lint/Deprecation errors and warnings that came up

Other than, of course, updating the demo apps of those libs mentioned above?

**Comment by ParaskP7** [comment]
👋 @thomashorta , yes the above course of actions, including updating the demo apps, seems good to me! 🎯

FYI: At some point in the future, we should also consider removing the `targetSdkVersion` configuration from all the internal libs of ours, that is, on any of their library module(s), while still keeping that on any of the demo application module(s). But still, even when we do so, it would be good to first update the `compileSdkVersion` on all the internal libs we use, in order to prepare our main JP/WPAndroid repo for the final `targetSdkVersion` upgrade, and this way stay on the safe side that we've done everything we could to prepare for the upgrade.

**Comment by thomashorta** [comment]
I created a new issue to track the `compileSdkVersion` updates to our libs.
https://github.com/wordpress-mobile/WordPress-Android/issues/19106

**Comment by thomashorta** [comment]
I'm closing this PR as complete without doing the changes in stories and FluxC sample apps, because they are just samples apps and we should always grant full permission to media when using them for tests. For instance: the `stories` sample app doesn't even request the permissions in runtime, so we need to grant the permissions through the system app settings.
